### DISCUSSION TITLE
Maintain cursor positon on Enter+Leave Insert mode

### DISCUSSION THREAD

**Discussion by eugenesvk** [NeoVintageous/NeoVintageous#935] [Ideas]
Currently when you enter and exit insert mode the cursor moves by 1 to the left, that's very unexpected
Is there a way to switch this to maintain cursor position?

**Comment by gerardroche** [comment]
This is the same in Vim. I don't know what the reasoning is, but no doubt there is one.

**Comment by eugenesvk** [comment]
> This is the same in Vim. I don't know what the reasoning is, but no doubt there is one.

Sure there is one, it's just not a good one

https://vimhelp.org/vim_faq.txt.html#faq-10.2

> No. It is not possible to change this behavior. The cursor is always positioned on a valid character (unless you have virtual-edit mode enabled). So, if you are appending text to the end of a line, when you return to command mode the cursor must drop back onto the last character you typed. For consistency sake, the cursor drops back everywhere, even if you are in the middle of a line.

First, it's perfectly possible to rest on the valid newline character like some other text editors do
Second, consistency with the more rare case of a line end is less desirable vs. the most basic intuition that enter/exit should not change anything, it's better for the line end to be special

But I'm not arguing for a change in the bad defaults, I'd just like to be able to configure them to be better

**Comment by gerardroche** [comment]
It has never bothered me. It's not something I want to add, because I think it's fine as it is. I bow to the Vim devs.

**Reply by sourcevault** [reply]
It's not a good idea to go against the Vim Gandalfs, there in lies the fire and chaos of mordor.

![giphy](https://github.com/NeoVintageous/NeoVintageous/assets/20347647/b1517422-fcc5-48cd-889d-ce9384da9ac1)

**Reply by eugenesvk** [reply]
Could you at least point to the code part that does this so I could override it locally?

**Reply by gerardroche** [reply]
It depends on what you're looking for: a quick hack just to make it work or a robust solution.

As a quick hack I would always go with an custom event outside the plugin to see if that will get me what I what I want.

This kinda works and might be what you want. None of this stuff is a public api so it will break when things are renamed, changed, removed. But you can always check the changelog for what changed.

```
import sublime_plugin

class AdjustCursorListener(sublime_plugin.EventListener):

    def on_post_text_command(self, view, command_name, args):
        if command_name == 'nv_enter_normal_mode':
            if 'mode' in args and args['mode'] != 'mode_normal':
                view.run_command('adjust_cursor')

class AdjustCursorCommand(sublime_plugin.TextCommand):

    def run(self, edit):
        point = self.view.sel()[0].b
        self.view.sel().clear()
        self.view.sel().add(point + 1)
```

You might not like the fact that now entering normal mode after pressing **a** keeps the cursor at the same point. 

I don't think the code above will work in all cases and I'd have to test it to see what side effects, if any, it has.

In the code you do the same by removing lines 802 and 803 here:

https://github.com/NeoVintageous/NeoVintageous/blob/4213c4cf3b37a3da662b087dbb4c73ca6129a8ff/nv/commands.py#L802

The lines:

```
                if view.line(s.b).a != s.b:
                    return Region(s.b - 1)
```

As you can see, it moves the cursor back 1 character.

**Reply by eugenesvk** [reply]
Thanks, the last code change is much simpler and appending seems to work (move the cursor right) no mode change

**Reply by sourcevault** [reply]
I would go with patching the original code, I have made the changes that @eugenesvk requested on my system  and it seems okay ( even if its not vim default ).

I propose adding it as a configuration option to satisfy both "to the letter" vimmer and ones who want a different type of default.

**Reply by eugenesvk** [reply]
By the way, is there a similarly simple tweak to change the line ending behavior to make the cursor stay on the line ending character, i.e., +1 to the right of the last visible text character? This would allow the <kbd>␈</kbd> deleting behavior to be consistent everywhere, from the end and in the middle

Interestingly, when I jump to the end of the line, the cursor gets to the +1 place, but then any movement or delets makes it jump back (there is no space after `l`)

https://github.com/NeoVintageous/NeoVintageous/assets/12956286/23926807-3565-4332-bf12-b0c2f831f3d6



P.S.
Meanwhile I've added a config like that
https://github.com/eugenesvk/NeoVintageous/commit/6e6ec57b6cf97c7dce77e7045e7f54f858418739

```jsonc
// Edit behavior
"edit"                      	: { //|default|
  "move_left_on_insert_exit"	: false, //|true| move cursor left on ⓘ→Ⓝ to be consistent with line-end editing https://vimhelp.org/vim_faq.txt.html#faq-10.2
},
```

**Reply by gerardroche** [reply]
I have no problem adding settings for isolated tweaks here and there. The problem is that sometimes these things look simple but turn out to be more complicated that you first thought. It's a slippery slope into maintenance hell. The thing that concerns me about this particular tweak is that Vim hasn't provided an option, which tells me there are probably other reasons we are not aware of yet.

If all this tweak needs is the small change above then of course that will be option. Keep testing it to see if there are side effects or other issues. 

The line ending issue is a known issue e.g., see https://github.com/sublimehq/sublime_text/issues/2120 and https://github.com/sublimehq/sublime_text/issues/2121#issuecomment-377910324, I'm not sure if it is possible to keep the cursor on that eol character. It's a bit complicated.

**Reply by eugenesvk** [reply]
> The line ending issue is a known issue e.g., see [sublimehq/sublime_text#2120](https://github.com/sublimehq/sublime_text/issues/2120) and [sublimehq/sublime_text#2121 (comment)](https://github.com/sublimehq/sublime_text/issues/2121#issuecomment-377910324), I'm not sure if it is possible to keep the cursor on that eol character. It's a bit complicated

how does the plugin do it in the selection mode? That seems like the behavior I mentioned (and backspace works nicely by deleting the newline), so how to use it universally in the normal mode as well, not only in the seleciton mode?

**Reply by gerardroche** [reply]
What are the delete inconsistencies? I think selections let you select the eol `\n`. I can't remember if it's possible in normal mode.

**Reply by eugenesvk** [reply]
Yes, selections let you select the eol, that's what I meant, so it's already implemented somehow in NV!!! Is there another simple `-1` at the end of a line somewhere in the normal mode :)?

delete is inconsistent in a way that you can't delete (in the same way) the last character since for that you'd need to be to the right of that char, at the eol newline char (you can't delete `l` in the example above outside of that "bug")

**Reply by gerardroche** [reply]
I don't know if the block cursor can position itself on the eol. Read through the issues I posted and the referenced issues in those threads. There is probably a solution in there somewhere.

Pressing `x` i.e., delete character under cursor, should delete the character?

**Reply by eugenesvk** [reply]
> Pressing `x` i.e., delete character under cursor, should delete the character?

This is inconsistent, you need to use a different key

>I don't know if the block cursor can position itself on the eol. Read through the issues I posted and the referenced issues in those threads. There is probably a solution in there somewhere.

I have, they imply it's not possible consistently :( like in this still opened mouse bug https://github.com/sublimehq/sublime_text/issues/1469

But then `{"keys":["ctrl+right"        ],	"command":"move_to",	"args":{"to":"eol",         	                 "extend":false}},` works in NV normal mode
And cursor in selection mode works
So I'm left confused whether it's another x-years long ST bug or maybe it's somehow fixable now

**Reply by gerardroche** [reply]
What key are you trying  to use? In normal mode pressing x is consistent. To delete in insert mode press `a<bs>`. If you want to delete it from normal mode with `<bs>` map it. `nnoremap <bs> x` should do it.

**Reply by eugenesvk** [reply]
But I don't want to delete the char under cursor, but the char left from cursor so that it's consistent with deleting in the insert mode - always to the left of the visible cursor
So I'm binding to 
`["x]X        	🗘	delete N characters before the cursor [into register x]`

And this doesn't work at the end of the line

**Reply by gerardroche** [reply]
```
nnoremap <bs> hxl
```

**Reply by eugenesvk** [reply]
this trades breaking at the end of the line for breaking at the beginning of the line (and it's worse since deleting the first char weirdly jumps to the 3rd (which becomes 2nd) instead of staying on the 1st position)

**Reply by gerardroche** [reply]
Can Vim work the way you want?

**Reply by eugenesvk** [reply]
`virtualedit=onemore` works that way with `X` at the end of the line, it moves 1 virtual char past the end of the line

**Reply by gerardroche** [reply]
X works correctly. It's the same as dh. There's no simple fix, that I know, to emulate virtualedit=onemore. It's  certainly possible though.

**Reply by eugenesvk** [reply]
Seems like there is a simple `-1` fix in the cursor move command

https://github.com/eugenesvk/NeoVintageous/commit/e27f2b8477aced77731d1c19bdb096077fabc4b2

And also you need to disable the special function breaking it: `fix_eol_cursor` otherwise it tries to "restore" position, leading to deletes at line end jumping by two chars
https://github.com/eugenesvk/NeoVintageous/commit/22243a1de72b59b225e3730e8a8d7d0264296a8e

(hopefully nothing in the weeds will be broken much due to this)
